File: generate-search-index.js
const fs = require('fs');
const path = require('path');
const cheerio = require('cheerio');

// Configuration
const TOPICS_DIR = path.join(__dirname, 'topics');
const OUTPUT_FILE = path.join(__dirname, 'search-index.json');

// Unit code mapping from file names
const UNIT_MAPPING = {
    'a1-communication.html': { code: 'A.1', name: 'Communication' },
    'a2-hydration-nutrition.html': { code: 'A.2', name: 'Hydration and nutrition' },
    'a3-response.html': { code: 'A.3', name: 'Response' },
    'b1-generating-movement.html': { code: 'B.1', name: 'Generating movement in the body' },
    'b2-forces-motion.html': { code: 'B.2', name: 'Forces, motion and movement' },
    'b3-injury.html': { code: 'B.3', name: 'Injury' },
    'c1-individual-differences.html': { code: 'C.1', name: 'Individual differences' },
    'c2-motor-learning.html': { code: 'C.2', name: 'Motor learning' },
    'c3-motivation.html': { code: 'C.3', name: 'Motivation' },
    'c4-stress-coping.html': { code: 'C.4', name: 'Stress and coping' },
    'c5-psychological-skills.html': { code: 'C.5', name: 'Psychological skills' }
};

// Helper function to generate a slug from text
function generateSlug(text) {
    return text
        .toLowerCase()
        .replace(/[^\w\s-]/g, '') // Remove special characters
        .replace(/\s+/g, '-')      // Replace spaces with hyphens
        .replace(/-+/g, '-')       // Replace multiple hyphens with single
        .trim();
}

// Helper function to ensure unique ID
function ensureUniqueId(baseId, existingIds) {
    let id = baseId;
    let counter = 1;
    while (existingIds.has(id)) {
        id = `${baseId}-${counter}`;
        counter++;
    }
    existingIds.add(id);
    return id;
}

// Extract searchable content from an HTML file
function extractContent(filePath, fileName) {
    const html = fs.readFileSync(filePath, 'utf-8');
    const $ = cheerio.load(html);
    const searchEntries = [];
    const existingIds = new Set();
    let modified = false;

    // Get unit info
    const unitInfo = UNIT_MAPPING[fileName] || { code: 'General', name: '' };
    const relativePath = `topics/${fileName}`;

    // Track all existing IDs to avoid duplicates
    $('[id]').each((i, el) => {
        existingIds.add($(el).attr('id'));
    });

    // Extract H1 (page title)
    $('h1').each((i, el) => {
        const title = $(el).text().trim();
        if (title) {
            let id = $(el).attr('id');
            if (!id) {
                id = ensureUniqueId(generateSlug(title), existingIds);
                $(el).attr('id', id);
                modified = true;
            }
            searchEntries.push({
                title: title,
                unit: unitInfo.code,
                unitName: unitInfo.name,
                page: relativePath,
                anchor: `#${id}`,
                type: 'page-title'
            });
        }
    });

    // Extract H2 headings (major sections)
    $('h2').each((i, el) => {
        const title = $(el).text().trim();
        if (title) {
            let id = $(el).attr('id');
            if (!id) {
                id = ensureUniqueId(generateSlug(title), existingIds);
                $(el).attr('id', id);
                modified = true;
            }
            searchEntries.push({
                title: title,
                unit: unitInfo.code,
                unitName: unitInfo.name,
                page: relativePath,
                anchor: `#${id}`,
                type: 'section'
            });
        }
    });

    // Extract H3 headings (subsections and concepts)
    $('h3').each((i, el) => {
        const title = $(el).text().trim();
        if (title) {
            let id = $(el).attr('id');
            if (!id) {
                id = ensureUniqueId(generateSlug(title), existingIds);
                $(el).attr('id', id);
                modified = true;
            }
            searchEntries.push({
                title: title,
                unit: unitInfo.code,
                unitName: unitInfo.name,
                page: relativePath,
                anchor: `#${id}`,
                type: 'concept'
            });
        }
    });

    // Extract H4 headings (specific topics)
    $('h4').each((i, el) => {
        const title = $(el).text().trim();
        if (title) {
            let id = $(el).attr('id');
            if (!id) {
                id = ensureUniqueId(generateSlug(title), existingIds);
                $(el).attr('id', id);
                modified = true;
            }
            searchEntries.push({
                title: title,
                unit: unitInfo.code,
                unitName: unitInfo.name,
                page: relativePath,
                anchor: `#${id}`,
                type: 'topic'
            });
        }
    });

    // Extract key term titles
    $('.term-title').each((i, el) => {
        const title = $(el).text().trim();
        if (title) {
            // Find the parent card and ensure it has an ID
            const parent = $(el).closest('.key-term-card');
            let id = parent.attr('id');
            if (!id) {
                id = ensureUniqueId(`term-${generateSlug(title)}`, existingIds);
                parent.attr('id', id);
                modified = true;
            }
            searchEntries.push({
                title: title,
                unit: unitInfo.code,
                unitName: unitInfo.name,
                page: relativePath,
                anchor: `#${id}`,
                type: 'term'
            });
        }
    });

    // If we modified the HTML, write it back
    if (modified) {
        fs.writeFileSync(filePath, $.html(), 'utf-8');
        console.log(`✓ Updated ${fileName} with missing IDs`);
    }

    return searchEntries;
}

// Main function
function generateSearchIndex() {
    console.log('Generating search index...\n');

    let allEntries = [];

    // Process all topic files
    const files = fs.readdirSync(TOPICS_DIR);
    files.forEach(fileName => {
        if (fileName.endsWith('.html') && UNIT_MAPPING[fileName]) {
            const filePath = path.join(TOPICS_DIR, fileName);
            console.log(`Processing ${fileName}...`);
            const entries = extractContent(filePath, fileName);
            allEntries = allEntries.concat(entries);
            console.log(`  Found ${entries.length} searchable items`);
        }
    });

    // Write the search index
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(allEntries, null, 2), 'utf-8');

    console.log(`\n✓ Search index generated successfully!`);
    console.log(`  Total entries: ${allEntries.length}`);
    console.log(`  Output: ${OUTPUT_FILE}`);
}

// Run the script
try {
    generateSearchIndex();
} catch (error) {
    console.error('Error generating search index:', error);
    process.exit(1);
}

File: generate-search-index.py
#!/usr/bin/env python3
"""
Generate search index for IB SEHS Study Guides
Scans HTML files and creates a searchable JSON index
"""

import json
import os
import re
from html.parser import HTMLParser
from pathlib import Path

# Unit code mapping from file names
UNIT_MAPPING = {
    'a1-communication.html': {'code': 'A.1', 'name': 'Communication'},
    'a2-hydration-nutrition.html': {'code': 'A.2', 'name': 'Hydration and nutrition'},
    'a3-response.html': {'code': 'A.3', 'name': 'Response'},
    'b1-generating-movement.html': {'code': 'B.1', 'name': 'Generating movement in the body'},
    'b2-forces-motion.html': {'code': 'B.2', 'name': 'Forces, motion and movement'},
    'b3-injury.html': {'code': 'B.3', 'name': 'Injury'},
    'c1-individual-differences.html': {'code': 'C.1', 'name': 'Individual differences'},
    'c2-motor-learning.html': {'code': 'C.2', 'name': 'Motor learning'},
    'c3-motivation.html': {'code': 'C.3', 'name': 'Motivation'},
    'c4-stress-coping.html': {'code': 'C.4', 'name': 'Stress and coping'},
    'c5-psychological-skills.html': {'code': 'C.5', 'name': 'Psychological skills'}
}


class HeadingParser(HTMLParser):
    """Parse HTML and extract headings with their IDs"""
    
    def __init__(self):
        super().__init__()
        self.headings = []
        self.term_titles = []
        self.current_tag = None
        self.current_attrs = {}
        self.current_content = []
        self.in_term_title = False
        self.all_ids = set()
        
    def handle_starttag(self, tag, attrs):
        attrs_dict = dict(attrs)
        
        if tag in ['h1', 'h2', 'h3', 'h4']:
            self.current_tag = tag
            self.current_attrs = attrs_dict
            self.current_content = []
        elif tag == 'div' and attrs_dict.get('class') == 'term-title':
            self.in_term_title = True
            self.current_content = []
            # Store parent attributes if available
            self.current_attrs = attrs_dict
            
        # Track all existing IDs
        if 'id' in attrs_dict:
            self.all_ids.add(attrs_dict['id'])
    
    def handle_data(self, data):
        if self.current_tag or self.in_term_title:
            self.current_content.append(data)
    
    def handle_endtag(self, tag):
        if tag in ['h1', 'h2', 'h3', 'h4'] and self.current_tag == tag:
            content = ''.join(self.current_content).strip()
            if content:
                self.headings.append({
                    'tag': tag,
                    'content': content,
                    'id': self.current_attrs.get('id', '')
                })
            self.current_tag = None
            self.current_content = []
        elif tag == 'div' and self.in_term_title:
            content = ''.join(self.current_content).strip()
            if content:
                self.term_titles.append(content)
            self.in_term_title = False
            self.current_content = []


def generate_slug(text):
    """Generate a URL-friendly slug from text"""
    # Convert to lowercase
    slug = text.lower()
    # Remove special characters except spaces and hyphens
    slug = re.sub(r'[^\w\s-]', '', slug)
    # Replace spaces with hyphens
    slug = re.sub(r'\s+', '-', slug)
    # Remove multiple hyphens
    slug = re.sub(r'-+', '-', slug)
    return slug.strip('-')


def ensure_unique_id(base_id, existing_ids):
    """Ensure ID is unique by adding counter if needed"""
    id_val = base_id
    counter = 1
    while id_val in existing_ids:
        id_val = f"{base_id}-{counter}"
        counter += 1
    existing_ids.add(id_val)
    return id_val


def add_ids_to_html(html_content, headings, existing_ids):
    """Add missing IDs to headings in HTML content"""
    modified = False
    
    for heading in headings:
        if not heading['id']:
            # Generate unique ID
            base_id = generate_slug(heading['content'])
            unique_id = ensure_unique_id(base_id, existing_ids)
            heading['id'] = unique_id
            
            # Find and replace the heading tag to add ID
            tag = heading['tag']
            content = heading['content']
            
            # Pattern to match the opening tag
            pattern = f'<{tag}([^>]*)>({re.escape(content)})</{tag}>'
            
            def replacer(match):
                attrs = match.group(1)
                # Only add ID if it's not already there
                if 'id=' not in attrs:
                    if attrs.strip():
                        new_tag = f'<{tag}{attrs} id="{unique_id}">{match.group(2)}</{tag}>'
                    else:
                        new_tag = f'<{tag} id="{unique_id}">{match.group(2)}</{tag}>'
                    return new_tag
                return match.group(0)
            
            new_content = re.sub(pattern, replacer, html_content, count=1)
            if new_content != html_content:
                html_content = new_content
                modified = True
    
    return html_content, modified


def extract_content(file_path, file_name):
    """Extract searchable content from an HTML file"""
    with open(file_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # Parse HTML
    parser = HeadingParser()
    parser.feed(html_content)
    
    unit_info = UNIT_MAPPING.get(file_name, {'code': 'General', 'name': ''})
    relative_path = f"topics/{file_name}"
    
    search_entries = []
    existing_ids = parser.all_ids.copy()
    
    # Process headings and add missing IDs
    html_content, modified = add_ids_to_html(html_content, parser.headings, existing_ids)
    
    # Create search entries
    for heading in parser.headings:
        heading_id = heading['id']
        
        # Determine type based on tag
        type_map = {
            'h1': 'page-title',
            'h2': 'section',
            'h3': 'concept',
            'h4': 'topic'
        }
        
        search_entries.append({
            'title': heading['content'],
            'unit': unit_info['code'],
            'unitName': unit_info['name'],
            'page': relative_path,
            'anchor': f"#{heading_id}",
            'type': type_map.get(heading['tag'], 'section')
        })
    
    # Add term titles
    for term in parser.term_titles:
        term_id = ensure_unique_id(f"term-{generate_slug(term)}", existing_ids)
        search_entries.append({
            'title': term,
            'unit': unit_info['code'],
            'unitName': unit_info['name'],
            'page': relative_path,
            'anchor': f"#{term_id}",
            'type': 'term'
        })
    
    # Write back if modified
    if modified:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(html_content)
        print(f"✓ Updated {file_name} with missing IDs")
    
    return search_entries


def main():
    """Main function to generate search index"""
    print("Generating search index...\n")
    
    script_dir = Path(__file__).parent
    topics_dir = script_dir / 'topics'
    output_file = script_dir / 'search-index.json'
    
    all_entries = []
    
    # Process all topic files
    for file_name in UNIT_MAPPING.keys():
        file_path = topics_dir / file_name
        if file_path.exists():
            print(f"Processing {file_name}...")
            try:
                entries = extract_content(file_path, file_name)
                all_entries.extend(entries)
                print(f"  Found {len(entries)} searchable items")
            except Exception as e:
                print(f"  Error processing {file_name}: {e}")
    
    # Write search index
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_entries, f, indent=2, ensure_ascii=False)
    
    print(f"\n✓ Search index generated successfully!")
    print(f"  Total entries: {len(all_entries)}")
    print(f"  Output: {output_file}")


if __name__ == '__main__':
    main()
